### PR TITLE
Added GraphQL Queries, Basic Functionalities

### PR DESCRIPTION
**Pull request by Dutta-SD**
#2 
Added GraphQL Query to Fetch Repo Stats.

### CODE REVIEW

File: Get_Repo_Details/config.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+## Configuration constants
+import os
+
+BASE_URL = "https://api.github.com/graphql" # Github base url
+GITHUB_TOKEN = os.environ['GITHUB_TOKEN']  # Github token environment variable
+REPONAME, USERNAME = None, None
```
Comment (smaranjitghose) [inline Get_Repo_Details/config.py]: this is supposed to be entered as a single string

File: Get_Repo_Details/config.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (smaranjitghose) [inline Get_Repo_Details/config.py]: use fire library to take input

File: Get_Repo_Details/get_repo_details.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# Script to GET Repo details from user
+# @Author - Sandip Dutta
+# Email - duttasandip11100@gmail.com
+
+# IMPORTS
+import json
+import requests
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py:3]: Have this in a single script. At this stage, there is no need for modularization

File: Get_Repo_Details/get_repo_details.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import config
+
+def getRepoDetails():
+  # Main function
+  # Set names of Repo and gets data from API Query
+
+  config.REPONAME = input("Enter Repo Name as it appears : ")
+  config.USERNAME = input("Enter the owner of the repository as it appears : ")
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: All this needs to be taken using command line input

File: Get_Repo_Details/get_repo_details.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  return num_collaborators, num_pullRequests, num_stars, num_commits
+
+
+if __name__ == '__main__':
+  num_collaborators, num_pullRequests, num_stars, num_commits = getRepoDetails()
+
+  # Pretty display 
+  print("-" * 80)
+  print(f"Stats for Repository {config.REPONAME} (Owner : {config.USERNAME})")
+  print(f"Number of collaborators : {num_collaborators}")
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: Use pprint instead of print

**Review by smaranjitghose — CHANGES_REQUESTED** [review]

**Comment by smaranjitghose** [comment]
Add a snapshot of the terminal output to your PR as well

File: Get_Repo_Details/get_repo_details.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# Script to GET Repo details from user
+# @Author - Sandip Dutta
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: We will have all of the contributors and maintainers on the documentation page. This is not required in the scripts as of now

File: Get_Repo_Details/config.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+def getQuery():
+  # Reuturns the Query String
```
Comment (smaranjitghose) [inline Get_Repo_Details/config.py]: Use DocStrings for describing functions

**Review by smaranjitghose — CHANGES_REQUESTED** [review]

File: Get_Repo_Details/get_repo_details.py (hunk 6 above)
Comment (Dutta-SD) [reply Get_Repo_Details/get_repo_details.py]: Yes I will remove these lines from the script.

File: Get_Repo_Details/get_repo_details.py (hunk 3 above)
Comment (Dutta-SD) [reply Get_Repo_Details/get_repo_details.py:3]: Yes. Converted this into a single script

**Comment by Dutta-SD** [comment]
Hi, This is the screenshot of the app.
Git_Get_repo_Details

UPDATE - This was before adding and using fire module.

File: Get_Repo_Details/get_repo_details.py (hunk 5 above)
Comment (Dutta-SD) [reply Get_Repo_Details/get_repo_details.py]: Should I print the whole returned json response using pprint ? Or print this individually  ?

File: Get_Repo_Details/config.py (hunk 1 above)
Comment (Dutta-SD) [reply Get_Repo_Details/config.py]: Okay I made this into a command line tool using fire. Is it okay?

Comment (anushbhatia) [reply Get_Repo_Details/config.py]: ya

File: Get_Repo_Details/get_repo_details.py (hunk 5 above)
Comment (anushbhatia) [reply Get_Repo_Details/get_repo_details.py]: print complete json

**Comment by anushbhatia** [comment]
Please use fire to get command line arguments as inputs properly. Your current script lack's it.

File: Get_Repo_Details/get_repo_details.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        totalCount
+      }}
+      stargazers {{
+        totalCount
+      }}
+      object(expression: "master") {{
+        ... on Commit {{
+          history {{
+            totalCount
+          }}
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: This is a totally unnecessary function definition so initialize a string. remove it and  keep it as a simple variable

File: Get_Repo_Details/get_repo_details.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  }}""".format(repoName, userName)   # GraphQL Query
+  return query
+
+
+def getRepoDetails(repoName, userName):
+  # Main function
+  # Set names of Repo and gets data from API Query
+
+  # Sanity check
+  assert (repoName is not None and userName is not None), "Invalid Details"
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: Not at all necessary. A simple exit code does the work

File: Get_Repo_Details/get_repo_details.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+  # Specify Headers
+  # Note - Can only see collaborators if it is own repository
+  headers = {'Authorization': "Token " + GITHUB_TOKEN}
+  # API Request
+  response = requests.post(
+    BASE_URL,
+    json={'query': getQuery(repoName, userName)},
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: # GitHub's API endpoint for  GraphQL
url = 'https://api.github.com/graphql'

# Headers for Authentication
headers = {'Authorization': "Token " + access_token}

# Making our API request
r = requests.post(url, json={'query': query}, headers=headers)

# Printing the Status Code to check for any error
print(r.status_code)

File: Get_Repo_Details/get_repo_details.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    headers=headers)
+  
+  json_data = json.loads(response.text)
+
+  # Extract meaningful Information from json data
+  # As per Query Structure
+  num_collaborators = json_data['data']['repository']['collaborators']['totalCount']
+  num_pullRequests = json_data['data']['repository']['pullRequests']['totalCount']
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py]: Use simple names

pr_count
colab_count
star_count
commit_count

File: Get_Repo_Details/get_repo_details.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  print("-" * 80)
+  print(f"Stats for Repository {repoName} (Owner : {userName})")
+  print(f"Number of collaborators : {num_collaborators}")
+  print(f"Number of Pull Requests : {num_pullRequests}")
+  print(f"Number of Stars : {num_stars}")
+  print(f"Number of Commits : {num_commits}")
+
+
+
+if __name__ == '__main__':
```
Comment (smaranjitghose) [inline Get_Repo_Details/get_repo_details.py:72]: Learn to use Fire in an effective way:

```python
import fire
import os
import json
import requests
import numpy as np
import pandas as pd


class GitKundli(object):

# Declare the GitHub access tokens globally

  def repo_info(self, username, reponame):
    """All the relevant code including print statements"""



if __name__ == '__main__':
  fire.Fire(Calculator)
```

**Review by smaranjitghose — CHANGES_REQUESTED** [review]

File: Get_Repo_Details/get_repo_details.py (hunk 8 above)
Comment (Dutta-SD) [reply Get_Repo_Details/get_repo_details.py]: Removed it

**Comment by Dutta-SD** [comment]
### This is how the CLI looks.
I added all changes. As specified, I displayed the json data.
get_repo_sandip_dutta

File: Get_Repo_Details/get_repo_details.py (hunk 12 above)
Comment (Dutta-SD) [reply Get_Repo_Details/get_repo_details.py:72]: Yes sir, I modified the code like this. Please give me some feedback if it is okay now

**Review by smaranjitghose — APPROVED** [review]

**Pull request merged**